File: src/ios/FilePicker.m
#import "FilePicker.h"
#import <Cordova/CDV.h>

@implementation FilePicker

- (void)pickFiles:(CDVInvokedUrlCommand*)command {
    self.callbackId = command.callbackId;

    NSArray *types = @[@"public.data"];
    UIDocumentPickerViewController *documentPicker = [[UIDocumentPickerViewController alloc] initWithDocumentTypes:types inMode:UIDocumentPickerModeOpen];
    documentPicker.delegate = self;
    documentPicker.allowsMultipleSelection = YES;

    [self.viewController presentViewController:documentPicker animated:YES completion:nil];
}

- (void)documentPicker:(UIDocumentPickerViewController *)controller didPickDocumentsAtURLs:(NSArray<NSURL *> *)urls {
    NSMutableArray *resultArray = [NSMutableArray array];

    for (NSURL *url in urls) {
        NSMutableDictionary *fileData = [NSMutableDictionary dictionary];
        NSString *filename = [url lastPathComponent];
        NSString *mimeType = [self mimeTypeForPath:url];
        [fileData setObject:filename forKey:@"filename"];
        [fileData setObject:mimeType forKey:@"filemime"];
        [fileData setObject:[url path] forKey:@"filepath"];
        [resultArray addObject:fileData];
    }

    CDVPluginResult* result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsArray:resultArray];
    [self.commandDelegate sendPluginResult:result callbackId:self.callbackId];
}

- (NSString *)mimeTypeForPath:(NSURL *)url {
    NSString *pathExtension = [url pathExtension];
    CFStringRef UTI = UTTypeCreatePreferredIdentifierForTag(kUTTagClassFilenameExtension, (__bridge CFStringRef)pathExtension, NULL);
    NSString *mimeType = (__bridge_transfer NSString *)UTTypeCopyPreferredTagWithClass(UTI, kUTTagClassMIMEType);
    CFRelease(UTI);
    return mimeType ? mimeType : @"application/octet-stream";
}

@end

File: src/ios/FilePicker.swift
import Foundation
import UIKit
import Cordova

@objc(FilePicker)
class FilePicker: CDVPlugin {
    
    @objc(selectFiles:)
    func selectFiles(command: CDVInvokedUrlCommand) {
        let documentPicker = UIDocumentPickerViewController(documentTypes: ["public.item"], in: .import)
        documentPicker.delegate = self
        documentPicker.allowsMultipleSelection = true
        
        self.viewController.present(documentPicker, animated: true, completion: nil)
    }
}

extension FilePicker: UIDocumentPickerDelegate {
    
    func documentPicker(_ controller: UIDocumentPickerViewController, didPickDocumentsAt urls: [URL]) {
        var resultArray = [[String: Any]]()
        
        for url in urls {
            if let fileData = getFileMetadata(from: url) {
                resultArray.append(fileData)
            }
        }
        
        let pluginResult = CDVPluginResult(status: CDVCommandStatus_OK, messageAs: resultArray)
        self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
    }
    
    func documentPickerWasCancelled(_ controller: UIDocumentPickerViewController) {
        let pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: "File selection cancelled")
        self.commandDelegate.send(pluginResult, callbackId: command.callbackId)
    }
    
    private func getFileMetadata(from url: URL) -> [String: Any]? {
        var fileData: [String: Any] = [:]
        
        do {
            let resourceValues = try url.resourceValues(forKeys: [.nameKey, .typeIdentifierKey])
            fileData["filename"] = resourceValues.name
            fileData["filemime"] = resourceValues.typeIdentifier
            fileData["filepath"] = url.absoluteString
        } catch {
            print("Error fetching file metadata: \(error)")
            return nil
        }
        
        return fileData
    }
}
